sync/optimistic.md

```diff
@@ -0,0 +1,340 @@
+root of that block.
+
+Let `blocks: Dict[Root, BeaconBlock]` and `block_states: Dict[Root,
+BeaconState]` be the blocks (and accompanying states) that have been verified
+either completely or optimistically.
+
+Let `optimistic_roots: Set[Root]` be the set of `hash_tree_root(block)` for all
+optimistically imported blocks which have only received a `SYNCING` designation
+from an execution engine (i.e., they are not known to be `INVALID` or `VALID`).
+
+```python
+@dataclass
+class Store(object):
+    optimistic_roots: Set[Root]
+    head_block_root: Root
+    blocks: Dict[Root, BeaconBlock]
+    block_states: Dict[Root, BeaconState]
+```
+
+```python
+def is_optimistic(store: Store, block: BeaconBlock) -> bool:
+    return hash_tree_root(block) in store.optimistic_roots
+```
+
+```python
+def latest_verified_ancestor(store: Store, block: BeaconBlock) -> BeaconBlock:
+    # It is assumed that the `block` parameter is never an INVALID block.
+    while True:
+        if not is_optimistic(store, block) or block.parent_root == Root():
+            return block
+        block = store.blocks[block.parent_root]
+```
+
+```python
+def is_execution_block(block: BeaconBlock) -> BeaconBlock:
+    return block.body.execution_payload != ExecutionPayload()
+```
+
+```python
+def should_optimistically_import_block(store: Store, current_slot: Slot, block: BeaconBlock) -> bool:
+    justified_root = store.block_states[store.head_block_root].current_justified_checkpoint.root
+    justifed_is_verified = is_execution_block(store.blocks[justified_root])
+    block_is_deep = block.slot + SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY <= current_slot
+    return justified_is_verified or block_is_deep
+```
+
+Let only a node which returns `is_optimistic(store, head) is True` be an *optimistic
+node*. Let only a validator on an optimistic node be an *optimistic validator*.
+
+When this specification only defines behaviour for an optimistic
+node/validator, but *not* for the non-optimistic case, assume default
+behaviours without regard for optimistic sync.
+
+## Mechanisms
+
+### When to optimistically import blocks
+
+A block MAY be optimistically imported when
+`should_optimistically_import_block(store, current_slot, block)` returns
+`True`. This ensures that blocks are only optimistically imported if either:
+
+1. The justified checkpoint has execution enabled.
+1. The current slot (as per the system clock) is at least
+   `SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY` ahead of the slot of the block being
+   imported.
+
+*See [Fork Choice Poisoning](#fork-choice-poisoning) for the motivations behind
+these conditions.*
+
+### How to optimistically import blocks
+
+To optimistically import a block:
+
+- The [`execute_payload`](../specs/bellatrix/beacon-chain.md#execute_payload) function MUST return `True` if the execution
+	engine returns `SYNCING` or `VALID`. An `INVALID` response MUST return `False`.
+- The [`validate_merge_block`][] function MUST NOT raise an assertion if both the
```

Comment:
This means if these values  are unknown, bypass that and the rest of the validation. "MUST NOT raise an assertion" was not immediately clear to me.

maybe add "i.e. `if pow_block is one and pow_parent is None`, skip all assertions in `validate_merge_block`." or something to that tune

Reply:
Have we considered how to do this when `TERMINAL_BLOCK_HASH` is set?

Reply:
> Have we considered how to do this when TERMINAL_BLOCK_HASH is set?

Very good question. I considered this and added some helpful (but non-substantiative) changes in 50b236e5.

> maybe add "i.e. if pow_block is one and pow_parent is None, skip all assertions in validate_merge_block." or something to that tune

After considering `TERMINAL_BLOCK_HASH`, I think we do need *some* assertions in `validate_merge_block`. My changes in 50b236e5 might help to clarify what was not clear to you?

One option for making this clearer is to break `validate_merge_block` into two functions; one that checks `TERMINAL_BLOCK_HASH` things and another that calls out to the EL. Then, it would be simple for us to state that the EL function should always return some static value. I'm happy to try this if it's appealing to you.

Reply:
this should be fine